File: TinyScript/js/src/main/js/parser/ast/DeclareStmt.ts
import Stmt from "./Stmt";
import ASTNode from "./ASTNode";
import ASTNodeTypes from "./ASTNodeTypes";
import PeekTokenIterator from "../PeekTokenIterator";
import ParseException from "../ParseException";
import Expr from "./Expr";

export default class DeclareStmt extends Stmt {

    constructor() {
        super(ASTNodeTypes.DECLARE_STMT, "declare");
    }

    static parse(it: PeekTokenIterator): ASTNode {
        const {Factor} = require('./index')
        const stmt = new DeclareStmt();
        it.nextMatch("var");
        const peek = it.peek();
        const factor = Factor.parse(it);
        if (factor == null) {
            throw ParseException.fromToken(peek);
        }
        stmt.addChild(factor);
        const lexeme = it.nextMatch("=");
        const expr = Expr.parse(it);
        if (expr == null) {
            throw new ParseException("null exception")
        }
        stmt.addChild(expr);
        stmt.lexeme = lexeme;
        return stmt;
    }
}

File: TinyScript/java/src/main/java/paser/ast/DeclareStmt.java
package paser.ast;

import lexer.Token;
import paser.util.ParseException;
import paser.util.PeekTokenIterator;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/8/23 0:22
 **/
public class DeclareStmt extends Stmt {
    protected DeclareStmt(ASTNode parent) {
        super(parent, ASTNodeTypes.DECLARE_STMT, "declare");
    }


    public static ASTNode parse(ASTNode parent, PeekTokenIterator it) throws ParseException {
        DeclareStmt declareStmt = new DeclareStmt(parent);
        it.nextMatch("var");
        Token tkn = it.peek();
        ASTNode factor = Factor.parse(parent, it);
        if (factor == null) {
            throw new ParseException(tkn);
        }
        declareStmt.addChild(factor);
        Token token = it.nextMatch("=");
        ASTNode parse = Expr.parse(it);
        declareStmt.addChild(parse);
        declareStmt.setLexeme(token);
        return declareStmt;
    }
}
